File: sdk/javascript/gateway-sdk.js
/**
 * GateWay License SDK for JavaScript/Node.js
 * Version: 1.0.0
 */

class GatewaySDK {
  constructor(options = {}) {
    this.apiKey = options.apiKey
    this.baseUrl = options.baseUrl || 'https://your-project.supabase.co/functions/v1'
    this.timeout = options.timeout || 30000
    
    if (!this.apiKey) {
      throw new Error('API key is required')
    }
  }

  /**
   * Validate a license
   * @param {string} licenseKey - The license key to validate
   * @param {string} hwid - Hardware ID of the machine
   * @param {string} productId - Optional product ID for additional validation
   * @returns {Promise<Object>} Validation result
   */
  async validateLicense(licenseKey, hwid, productId = null) {
    try {
      const response = await this._makeRequest('/license-api/validate', 'POST', {
        license_key: licenseKey,
        hwid: hwid,
        product_id: productId
      })

      return response
    } catch (error) {
      throw new Error(`License validation failed: ${error.message}`)
    }
  }

  /**
   * Activate a license on current machine
   * @param {string} licenseKey - The license key to activate
   * @param {string} hwid - Hardware ID of the machine
   * @param {string} machineName - Optional machine name
   * @returns {Promise<Object>} Activation result
   */
  async activateLicense(licenseKey, hwid, machineName = null) {
    try {
      const response = await this._makeRequest('/license-api/activate', 'POST', {
        license_key: licenseKey,
        hwid: hwid,
        machine_name: machineName
      })

      return response
    } catch (error) {
      throw new Error(`License activation failed: ${error.message}`)
    }
  }

  /**
   * Deactivate a license from current machine
   * @param {string} licenseKey - The license key to deactivate
   * @param {string} hwid - Hardware ID of the machine
   * @returns {Promise<Object>} Deactivation result
   */
  async deactivateLicense(licenseKey, hwid) {
    try {
      const response = await this._makeRequest('/license-api/deactivate', 'POST', {
        license_key: licenseKey,
        hwid: hwid
      })

      return response
    } catch (error) {
      throw new Error(`License deactivation failed: ${error.message}`)
    }
  }

  /**
   * Get license information
   * @param {string} licenseKey - The license key
   * @returns {Promise<Object>} License information
   */
  async getLicenseInfo(licenseKey) {
    try {
      const response = await this._makeRequest(`/license-api/info?license_key=${licenseKey}`, 'GET')
      return response
    } catch (error) {
      throw new Error(`Failed to get license info: ${error.message}`)
    }
  }

  /**
   * Send heartbeat to keep license active
   * @param {string} licenseKey - The license key
   * @param {string} hwid - Hardware ID of the machine
   * @param {string} status - Current application status
   * @returns {Promise<Object>} Heartbeat result
   */
  async sendHeartbeat(licenseKey, hwid, status = 'running') {
    try {
      const response = await this._makeRequest('/license-api/heartbeat', 'POST', {
        license_key: licenseKey,
        hwid: hwid,
        status: status
      })

      return response
    } catch (error) {
      throw new Error(`Heartbeat failed: ${error.message}`)
    }
  }

  /**
   * Generate hardware ID for current machine
   * @returns {string} Hardware ID
   */
  generateHWID() {
    if (typeof window !== 'undefined') {
      // Browser environment
      const canvas = document.createElement('canvas')
      const ctx = canvas.getContext('2d')
      ctx.textBaseline = 'top'
      ctx.font = '14px Arial'
      ctx.fillText('Hardware fingerprint', 2, 2)
      
      const fingerprint = [
        navigator.userAgent,
        navigator.language,
        screen.width + 'x' + screen.height,
        new Date().getTimezoneOffset(),
        canvas.toDataURL()
      ].join('|')
      
      return this._hashString(fingerprint)
    } else {
      // Node.js environment
      const os = require('os')
      const crypto = require('crypto')
      
      const machineId = [
        os.hostname(),
        os.arch(),
        os.platform(),
        os.cpus()[0].model,
        Object.values(os.networkInterfaces())
          .flat()
          .find(i => !i.internal && i.mac !== '00:00:00:00:00:00')?.mac || 'unknown'
      ].join('|')
      
      return crypto.createHash('md5').update(machineId).digest('hex')
    }
  }

  /**
   * Start automatic heartbeat
   * @param {string} licenseKey - The license key
   * @param {string} hwid - Hardware ID
   * @param {number} interval - Heartbeat interval in milliseconds (default: 5 minutes)
   * @returns {number} Interval ID for stopping heartbeat
   */
  startHeartbeat(licenseKey, hwid, interval = 300000) {
    return setInterval(async () => {
      try {
        await this.sendHeartbeat(licenseKey, hwid)
      } catch (error) {
        console.error('Heartbeat error:', error.message)
      }
    }, interval)
  }

  /**
   * Stop automatic heartbeat
   * @param {number} intervalId - The interval ID returned by startHeartbeat
   */
  stopHeartbeat(intervalId) {
    clearInterval(intervalId)
  }

  /**
   * Make HTTP request to API
   * @private
   */
  async _makeRequest(endpoint, method, data = null) {
    const url = `${this.baseUrl}${endpoint}`
    
    const options = {
      method: method,
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': this.apiKey
      }
    }

    if (data && method !== 'GET') {
      options.body = JSON.stringify(data)
    }

    // Add timeout
    const controller = new AbortController()
    const timeoutId = setTimeout(() => controller.abort(), this.timeout)
    options.signal = controller.signal

    try {
      const response = await fetch(url, options)
      clearTimeout(timeoutId)
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`)
      }

      return await response.json()
    } catch (error) {
      clearTimeout(timeoutId)
      if (error.name === 'AbortError') {
        throw new Error('Request timeout')
      }
      throw error
    }
  }

  /**
   * Hash string using simple algorithm
   * @private
   */
  _hashString(str) {
    let hash = 0
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i)
      hash = ((hash << 5) - hash) + char
      hash = hash & hash // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(16)
  }
}

// Export for different environments
if (typeof module !== 'undefined' && module.exports) {
  module.exports = GatewaySDK
} else if (typeof window !== 'undefined') {
  window.GatewaySDK = GatewaySDK
}

File: sdk/python/gateway_sdk.py
"""
GateWay License SDK for Python
Version: 1.0.0
"""

import requests
import hashlib
import platform
import uuid
import time
import threading
from typing import Optional, Dict, Any

class GatewaySDK:
    def __init__(self, api_key: str, base_url: str = None, timeout: int = 30):
        """
        Initialize GateWay SDK
        
        Args:
            api_key (str): Your API key
            base_url (str): Base URL for API (optional)
            timeout (int): Request timeout in seconds
        """
        if not api_key:
            raise ValueError("API key is required")
            
        self.api_key = api_key
        self.base_url = base_url or "https://your-project.supabase.co/functions/v1"
        self.timeout = timeout
        self._heartbeat_thread = None
        self._heartbeat_stop = False
        
    def validate_license(self, license_key: str, hwid: str, product_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Validate a license
        
        Args:
            license_key (str): The license key to validate
            hwid (str): Hardware ID of the machine
            product_id (str, optional): Product ID for additional validation
            
        Returns:
            dict: Validation result
        """
        try:
            response = self._make_request('/license-api/validate', 'POST', {
                'license_key': license_key,
                'hwid': hwid,
                'product_id': product_id
            })
            return response
        except Exception as e:
            raise Exception(f"License validation failed: {str(e)}")
    
    def activate_license(self, license_key: str, hwid: str, machine_name: Optional[str] = None) -> Dict[str, Any]:
        """
        Activate a license on current machine
        
        Args:
            license_key (str): The license key to activate
            hwid (str): Hardware ID of the machine
            machine_name (str, optional): Machine name
            
        Returns:
            dict: Activation result
        """
        try:
            response = self._make_request('/license-api/activate', 'POST', {
                'license_key': license_key,
                'hwid': hwid,
                'machine_name': machine_name or platform.node()
            })
            return response
        except Exception as e:
            raise Exception(f"License activation failed: {str(e)}")
    
    def deactivate_license(self, license_key: str, hwid: str) -> Dict[str, Any]:
        """
        Deactivate a license from current machine
        
        Args:
            license_key (str): The license key to deactivate
            hwid (str): Hardware ID of the machine
            
        Returns:
            dict: Deactivation result
        """
        try:
            response = self._make_request('/license-api/deactivate', 'POST', {
                'license_key': license_key,
                'hwid': hwid
            })
            return response
        except Exception as e:
            raise Exception(f"License deactivation failed: {str(e)}")
    
    def get_license_info(self, license_key: str) -> Dict[str, Any]:
        """
        Get license information
        
        Args:
            license_key (str): The license key
            
        Returns:
            dict: License information
        """
        try:
            response = self._make_request(f'/license-api/info?license_key={license_key}', 'GET')
            return response
        except Exception as e:
            raise Exception(f"Failed to get license info: {str(e)}")
    
    def send_heartbeat(self, license_key: str, hwid: str, status: str = 'running') -> Dict[str, Any]:
        """
        Send heartbeat to keep license active
        
        Args:
            license_key (str): The license key
            hwid (str): Hardware ID of the machine
            status (str): Current application status
            
        Returns:
            dict: Heartbeat result
        """
        try:
            response = self._make_request('/license-api/heartbeat', 'POST', {
                'license_key': license_key,
                'hwid': hwid,
                'status': status
            })
            return response
        except Exception as e:
            raise Exception(f"Heartbeat failed: {str(e)}")
    
    def generate_hwid(self) -> str:
        """
        Generate hardware ID for current machine
        
        Returns:
            str: Hardware ID
        """
        try:
            # Get MAC address
            mac = ':'.join(['{:02x}'.format((uuid.getnode() >> elements) & 0xff) 
                           for elements in range(0, 2*6, 2)][::-1])
            
            # Combine machine identifiers
            machine_id = f"{platform.node()}-{platform.machine()}-{platform.processor()}-{mac}"
            
            # Create hash
            return hashlib.md5(machine_id.encode()).hexdigest()
        except Exception:
            # Fallback to UUID if hardware info fails
            return hashlib.md5(str(uuid.uuid4()).encode()).hexdigest()
    
    def start_heartbeat(self, license_key: str, hwid: str, interval: int = 300) -> None:
        """
        Start automatic heartbeat
        
        Args:
            license_key (str): The license key
            hwid (str): Hardware ID
            interval (int): Heartbeat interval in seconds (default: 5 minutes)
        """
        if self._heartbeat_thread and self._heartbeat_thread.is_alive():
            self.stop_heartbeat()
        
        self._heartbeat_stop = False
        self._heartbeat_thread = threading.Thread(
            target=self._heartbeat_worker,
            args=(license_key, hwid, interval),
            daemon=True
        )
        self._heartbeat_thread.start()
    
    def stop_heartbeat(self) -> None:
        """Stop automatic heartbeat"""
        self._heartbeat_stop = True
        if self._heartbeat_thread:
            self._heartbeat_thread.join(timeout=5)
    
    def _heartbeat_worker(self, license_key: str, hwid: str, interval: int) -> None:
        """Worker thread for heartbeat"""
        while not self._heartbeat_stop:
            try:
                self.send_heartbeat(license_key, hwid)
            except Exception as e:
                print(f"Heartbeat error: {e}")
            
            # Sleep in small intervals to allow quick stopping
            for _ in range(interval):
                if self._heartbeat_stop:
                    break
                time.sleep(1)
    
    def _make_request(self, endpoint: str, method: str, data: Optional[Dict] = None) -> Dict[str, Any]:
        """Make HTTP request to API"""
        url = f"{self.base_url}{endpoint}"
        
        headers = {
            'Content-Type': 'application/json',
            'x-api-key': self.api_key
        }
        
        try:
            if method == 'GET':
                response = requests.get(url, headers=headers, timeout=self.timeout)
            elif method == 'POST':
                response = requests.post(url, headers=headers, json=data, timeout=self.timeout)
            elif method == 'PUT':
                response = requests.put(url, headers=headers, json=data, timeout=self.timeout)
            elif method == 'DELETE':
                response = requests.delete(url, headers=headers, timeout=self.timeout)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.Timeout:
            raise Exception("Request timeout")
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request failed: {str(e)}")
        except ValueError as e:
            raise Exception(f"Invalid response: {str(e)}")

# Example usage
if __name__ == "__main__":
    # Initialize SDK
    sdk = GatewaySDK(api_key="your-api-key-here")
    
    # Generate HWID
    hwid = sdk.generate_hwid()
    print(f"Hardware ID: {hwid}")
    
    # Validate license
    try:
        result = sdk.validate_license("YOUR-LICENSE-KEY", hwid)
        if result.get('valid'):
            print("License is valid!")
            print(f"Product: {result['license']['product']['name']}")
        else:
            print(f"License invalid: {result.get('error')}")
    except Exception as e:
        print(f"Error: {e}")